### ISSUE TITLE
Window, viewer and layer API

### ISSUE THREAD

**Issue by royerloic** [napari/napari#4]
Here is a mockup for what the viewer and layer api should look like (roughly):

Create a window:
`window = gui.newwindow(name, ...other options...)`
Note: We should distinguish the concept of 'window' and 'viewer'.
you can have several viewers in a window, organised either as tabs
or side by side, etc... Tabs are probably the best way to do it.

Create or removes a viewer in a window:
`viewer = window.addviewer(name=None, ...other options...)` %optional name
`viewer = window.removeviewer()`

Most of the time we just want a viewer straight away:
`viewer = gui.newviewer()`
This creates the window on the fly...

Even more often, we want an image displayed in a viewer inside of a window, in a one-liner:
`viewer = gui.imshow(...list of numpy arrays..., rgb=None)`

semi-internal api for adding/removing a layer of a given type:
`viewer.addlayer(ImageLayer(... list of numpy arrays ...))`
`viewer.removelayer(name or index of layer)`

convenience api for common layers:
`viewer.addimages(... list of numpy arrays ...)`
`viewer.addpoints(... list of points as nparray ...)`
`viewer.addlines(... list of lines as nparray  ...)`
etc...

get a layer from a viewer:
`layer = viewer.getlayer(...index or name of layer...)`

set the viewer sliders on a given layer:
`layer.setposition( tuple of ints )`
Note this must be a complete tuple that also includes the displayed dims!
for simplicity we should also have:
`layer.setposition(dimindex,position )`

Different layers have different available properties, for example:
`layer.cmap = ...`
`layer.opacity = ...`
etc...
In the case of image viewers, we have an api for the different layout modes etc...

We can always retrieve the reference to the data that the layer represents:
`data = layer.data`
for example the numpy array, etc...

And force the layer to refresh itself and reflect any change of the data:
`layer.refresh()`
Note: we don't have yet a mechanism to know when numpy arrays have changed
their content for example...

**Comment by jni** [comment]
I have many comments. =P

1. I suggest we try to mimic the matplotlib (and related) APIs as much as possible. First things first: I don't want window creation to be (in general) a public facing thing. And we should make using the viewer *super super* quick. I don't want to have to create a window then add a viewer to that window then add layers to that viewer. I want to say `window = napari.show_slices(arr)`. This creates the window, viewer, etc. for me. The returned window object should then have accessible all the remaining objects in the hierarchy. (@royerloic I just saw upon rereading that you propose this... That should be front and center. I also think that we should not use `imshow`. It's a little too confusing with existing libraries. Let's come up with a name that has no baggage. I like `show_slices`, `ndshow`, `ndview`... 

2. Also, until we see that we really need it, I would like to keep these objects super simple, almost `dataclass`es. For example, instead of `window.addviewer`, I would prefer `window.viewers.append()`, where viewers is a good ol' Python list, as well as `viewer.layers.append()`

3. Maybe the above is not *the* public API, but it should be an option. Specifically, I want this because it will force us to keep all the objects simple. I *really* don't want to end up where Qt is, which is a nauseating, insanely deep inheritance hierarchy. 

4. Most specifically, `layer = viewer.getlayer(...index or name of layer...)` I really dislike: `arr = viewer.layers[index]` seems much more Pythonic to me.

5. I think a very cool feature would be viewer.link_to(otherviewer), where viewers of the same dimension end up linked. This means the sliders end up linked *and* which dimensions are sliders are also linked. That would be a killer feature imho.

6. I am undecided between `viewer.add_points(arr)`, `viewer.show(Points(arr))`, and `viewer.show_points(arr)`. The second one seems more transparently extensible. 

7. "And force the layer to refresh itself and reflect any change of the data: `layer.refresh()` Note: we don't have yet a mechanism to know when numpy arrays have changed their content for example..." I *think* there's a library called `traits` or `traitlets` that lets you deal with this, but I have never grokked it. But it seems like we should look into it for this. =)

At any rate, yes, having an issue where we look at Napari from an API perspective is super important, so thanks for bringing it up! =)

Something not addressed at all here: how do I say: apply plugin "whatever" to layers x, y, z? Actually this is a very critical question that also drives us to simplicity: if we want to be able to record "reasonable" Python when running analysis in Napari, we can't have the code look like

```
viewer.addlayer(plugins.skimage.apply_watershed(viewer.layers[0], viewer.layers[1]))
```

Instead, it should be

```
watershed = skimage.segmentation.watershed(image, seeds)
```

Which becomes harder and harder to get back down to as the layers get more and more complex.

Going back to my [mockup](https://github.com/Napari/napari-core/issues/23), how does each action there map to code?

Finally, regarding the windowed/tabbed/docked layout options, I also think that we should keep this stuff super simple to begin with. Although I understand the concern about window explosions in Fiji, I think we should only have three options: add a layer, add a tab, and add a viewer/window. And I'd maybe even forgo tabs in the beginning, with just the idea that maybe we want to add them later.

**Comment by jni** [comment]
Incidentally, the repo structure is already getting way too complex I think...

**Comment by royerloic** [comment]
Great feedback Juan, thanks.
Some comments:

1a) I agree with matplotlib, we should match existing conventions as long as they make sense to us. I am not sure that I like all aspects of matplotlib though. We should see on this goes item per item... 

1b) The API right now does not distinguish between private or public, right now we need to make sure that even the  eventually private API is well designed too. The window stuff is there anyway so I wanted to make it more elegant and possibly usefull to a wider audience. Of course, as you noticed, there is no need to deal with windows if you don't want, but you can if you want. Nick yesterday showed us his 11 Napari windows open and he mentioned that he would not be unhappy to have a bit more control... 

2) I am fine with that, as long as we can intercept the changes to teh kist to be able to trigger the side-effects required for managing the UI... @kne42 is that possible? I guess yes.

3) Yes. I believe in a solid, clear, and self-consistent core API with a flexible convenience layer that sits on top and provides shortcuts. You want both to get a good architecture that does not turn into a code salad and something that provides one liners for common code paths. In any case, the python-way of doing things should be pushed, so when relevant and appropriate we should go for functional interfaces, properties, annotations... I trust @kne42 and @jni  to keep us on the good path here.

4) I agree, much better. I like operators. 

5) Definitely!! I was also thinking about that! synced viewers are going to be very important, and we should have an easy  way to link that up. @jni : can you create an issue and suggest an API for it?

6) We need both, the one that takes the object is generic and supports the layer extension mechanism that we are setting up with @kne42 which lets anyone design new layer types and ho to display them. But we need convenience methods for the common layer types such as images, points, etc...

 7) We discussed this during the meeting, we decided to have refreshing done manually until we have a  solid and transparent way of doing it automatically. We can always have a property that controls this behaviour... We should definitely explore solutions for that. But the priority is quite low for now considering the more pressing features that need implementing...

8) Mockup: we should indeed revisit that in light of the code, and revisit the ode in light of the mockup.

9) 'Applying plugins' : That's a tough one that is going to require some creative thinking. One approach is to somehow intercept the creation of new arrays and append them as layers. But there is no way we can completely hide the calls to the viewer/layer API in the general case. I also think it is dangerous to start hiding things from users as this would lead to a mismatch between what the code says and what actually happens. It would also get very complicated, possibly brittle, and unmaintainable. 
 
I would suggest this for 'generated code' for each operation:
`
result = plugins.skimage.apply_watershed(viewer.layers[0], viewer.layers[1])
viewer.addlayer(result)
`

if the user generates the code to start with, they are free to do a bunch of computations,
and then have a bunch of calls to display it. Any additional processing done in the ui would
simply be appended...

9) Yes, window, viewer(possibly as tab), layer, this is what we are aiming for right now with @kne42 and @AhmetCanSolak .

**Comment by kne42** [comment]
> I suggest we try to mimic the matplotlib (and related) APIs as much as possible. 

I totally agree!

> I also think that we should not use `imshow`. It's a little too confusing with existing libraries. Let's come up with a name that has no baggage. I like `show_slices`, `ndshow`, `ndview`...

I think that it helps a lot with intuition. Since MPL is the standard, shouldn't we use `imshow`, wouldn't that be people's go-to for showing images?

> I would like to keep these objects super simple, almost `dataclass`es. For example, instead of `window.addviewer`, I would prefer `window.viewers.append()`, where viewers is a good ol' Python list, as well as `viewer.layers.append()`

Yeah, I like this idea as well! However, instead of traditional Python lists, we'd need to use classes based off of `collections.UserList` so that we can insert listeners for when the list updates. At least if this is the public-facing API.

> I _really_ don't want to end up where Qt is, which is a nauseating, insanely deep inheritance hierarchy.

Definitely. Simplicity & elegance is the best way to go.

> Most specifically, `layer = viewer.getlayer(...index or name of layer...)` I really dislike: `arr = viewer.layers[index]` seems much more Pythonic to me.

Agreed.

> I think a very cool feature would be viewer.link_to(otherviewer), where viewers of the same dimension end up linked. This means the sliders end up linked _and_ which dimensions are sliders are also linked. That would be a killer feature imho.

Sure, but idk how much of a priority it should be.

> I am undecided between `viewer.add_points(arr)`, `viewer.show(Points(arr))`, and `viewer.show_points(arr)`. The second one seems more transparently extensible.

I think that the idea behind this is that `viewer.show(Points(arr))` is how we're going to make everything extensible and `viewer.add_points(arr)` is a convenience method. I'd like to, using `__getattr__` and `__dir__`, dynamically add registered layer types to go under `viewer.add_*`.

> I _think_ there's a library called `traits` or `traitlets` that lets you deal with this, but I have never grokked it. But it seems like we should look into it for this. =)

Sweet! I'll look into this :)

> Something not addressed at all here: how do I say: apply plugin "whatever" to layers x, y, z? Actually this is a very critical question that also drives us to simplicity: if we want to be able to record "reasonable" Python when running analysis in Napari, we can't have the code look like
> 
> ```
> viewer.addlayer(plugins.skimage.apply_watershed(viewer.layers[0], viewer.layers[1]))
> ```
> Instead, it should be
> 
> ```
> watershed = skimage.segmentation.watershed(image, seeds)
> ```
> Which becomes harder and harder to get back down to as the layers get more and more complex.
> 
> Going back to my [mockup](https://github.com/Napari/napari-core/issues/23), how does each action there map to code?

I see what you're getting at here. Sometimes plugins will be forced to wrap actual library functionality so we can't map it 1-1 for registered functions to what functions are displayed in the autogenerated code. I'm actually not really sure what to do about this and I think we should open a new issue to discuss this topic.

**Comment by sofroniewn** [comment]
Hi all,
Really like the way this conversation is taking shape, and agree with a lot of the sentiment around intuitive pythonic api's. The fact right now that I can just call `gui.imshow(data)` and it pops up an image in a window was really compelling to a python novice that I showed napari to the other day. That low barrier to entry is really important.

**Comment by kne42** [comment]
XREF: https://github.com/Napari/napari-gui/pull/12#issuecomment-431540278

**Comment by sofroniewn** [comment]
this has been closed by #290
